bitbots_dynup:
  engine:
    engine_rate:
      type: int
      description: "Rate at which the engine is looped"
      default_value: 240

    end_pose:
      arm_extended_length:
        type: double
        description: "Length of the arm when extended at the end pose. Depends on walkready of walking."
        default_value: 0.4
        validation:
          bounds<>: [0.0, 5.0]
      foot_distance:
        type: double
        description: "End pose foot distance from one foot to the other. Depends on walkready of walking."
        default_value: 0.2
        validation:
          bounds<>: [0.0, 5.0]
      hand_walkready_pitch:
        type: double
        description: "End pose hand pitch. Depends on walkready of walking."
        default_value: 8.0
        validation:
          bounds<>: [0.0, 90.0]
      hand_walkready_height:
        type: double
        description: "End pose hand height. Depends on walkready of walking."
        default_value: -0.29
        validation:
          bounds<>: [-1.0, 1.0]
      trunk_height:
        type: double
        description: "End pose trunk height. Depends on walkready of walking."
        default_value: 0.38
        validation:
          bounds<>: [0.0, 1.0]
      trunk_pitch:
        type: double
        description: "End pose trunk pitch. Depends on walkready of walking."
        default_value: 0.2
        validation:
          bounds<>: [0.0, 1.0]
      trunk_x_final:
        type: double
        description: "End pose position of the trunk in x direction. Depends on walkready of walking."
        default_value: 0.001
        validation:
          bounds<>: [0.0, 1.0]
      arm_side_offset_back:
        type: double
        description: "End pose arm side offset back. Depends on walkready of walking."
        default_value: 0.148
        validation:
          bounds<>: [0.0, 5.0]

    dynup_back:
      arms_angle_back:
        type: double
        description: "When fallen to the back: angle of the arms"
        default_value: 120.36
        validation:
          bounds<>: [0.0, 180.0]
      com_shift_1:
        type: double
        description: "When fallen to the back: shift of the center of mass"
        default_value: 0.051
        validation:
          bounds<>: [0.0, 1.0]
      com_shift_2:
        type: double
        description: "When fallen to the back: shift of the center of mass"
        default_value: 0.0
        validation:
          bounds<>: [0.0, 1.0]
      foot_angle:
        type: double
        description: "When fallen to the back: angle of the foot"
        default_value: 51.76
        validation:
          bounds<>: [0.0, 90.0]
      hands_behind_back_x:
        type: double
        description: "When fallen to the back: hands behind back x"
        default_value: 0.162
        validation:
          bounds<>: [0.0, 1.0]
      hands_behind_back_z:
        type: double
        description: "When fallen to the back: hands behind back z"
        default_value: 0.183
        validation:
          bounds<>: [0.0, 1.0]
      leg_min_length_back:
        type: double
        description: "When fallen to the back: minimum length of the leg"
        default_value: 0.253
        validation:
          bounds<>: [0.0, 1.0]
      time_foot_ground_back:
        type: double
        description: "When fallen to the back: time the foot is on the ground"
        default_value: 0.536
        validation:
          bounds<>: [0.0, 1.0]
      time_full_squat_hands:
        type: double
        description: "When fallen to the back: time the hands are in full squat"
        default_value: 0.172
        validation:
          bounds<>: [0.0, 1.0]
      time_full_squat_legs:
        type: double
        description: "When fallen to the back: time the legs are in full squat"
        default_value: 0.196
        validation:
          bounds<>: [0.0, 1.0]
      time_legs_close:
        type: double
        description: "When fallen to the back: time the legs are close"
        default_value: 0.068
        validation:
          bounds<>: [0.0, 1.0]
      trunk_height_back:
        type: double
        description: "When fallen to the back: trunk height"
        default_value: 0.179
        validation:
          bounds<>: [0.179, 1.0]
      trunk_overshoot_angle_back:
        type: double
        description: "When fallen to the back: the angle the trunk overshoots"
        default_value: 5.95
        validation:
          bounds<>: [0.0, 10.0]
      wait_in_squat_back:
        type: double
        description: "When fallen to the back: time to wait in squat position in seconds"
        default_value: 1.0
        validation:
          bounds<>: [0.0, 10.0]

    dynup_front:
      arm_side_offset_front:
        type: double
        description: "When fallen to the front: offset of the arms to the side"
        default_value: 0.13
        validation:
          bounds<>: [0.0, 1.0]
      hands_pitch:
        type: double
        description: "When fallen to the front: pitch of the hands"
        default_value: -65.86
        validation:
          bounds<>: [-90.0, 90.0]
      leg_min_length_front:
        type: double
        description: "When fallen to the front: minimum length of the leg"
        default_value: 0.244
        validation:
          bounds<>: [0.0, 1.0]
      max_leg_angle:
        type: double
        description: "When fallen to the front: maximum angle of the leg"
        default_value: 71.71
        validation:
          bounds<>: [0.0, 90.0]
      time_foot_close:
        type: double
        description: "When fallen to the front: time the foot is close"
        default_value: 0.0
        validation:
          bounds<>: [0.0, 1.0]
      time_foot_ground_front:
        type: double
        description: "When fallen to the front: time the foot is on the ground"
        default_value: 0.132
        validation:
          bounds<>: [0.0, 1.0]
      time_hands_front:
        type: double
        description: "When fallen to the front: time the hands are in front"
        default_value: 0.3
        validation:
          bounds<>: [0.0, 1.0]
      time_hands_rotate:
        type: double
        description: "When fallen to the front: time the hands rotate"
        default_value: 0.3
        validation:
          bounds<>: [0.0, 1.0]
      time_hands_side:
        type: double
        description: "When fallen to the front: time the hands are to the side"
        default_value: 0.3
        validation:
          bounds<>: [0.0, 1.0]
      time_to_squat:
        type: double
        description: "When fallen to the front: time to squat"
        default_value: 0.924
        validation:
          bounds<>: [0.0, 1.0]
      time_torso_45:
        type: double
        description: "When fallen to the front: time the torso is at 45 degrees"
        default_value: 0.462
        validation:
          bounds<>: [0.0, 1.0]
      trunk_overshoot_angle_front:
        type: double
        description: "When fallen to the front: the angle the trunk overshoots"
        default_value: -5.0
        validation:
          bounds<>: [-90.0, 90.0]
      trunk_x_front:
        type: double
        description: "When fallen to the front: trunk x position"
        default_value: 0.091
        validation:
          bounds<>: [0.0, 0.091]
      wait_in_squat_front:
        type: double
        description: "When fallen to the front: time to wait in squat position in seconds"
        default_value: 1.2
        validation:
          bounds<>: [0.0, 10.0]

    rise:
      rise_time:
        type: double
        description: "Time to rise in seconds"
        default_value: 2.0
        validation:
          bounds<>: [0.0, 10.0]

    descend:
      descend_time:
        type: double
        description: "Time to descend in seconds"
        default_value: 0.25
        validation:
          bounds<>: [0.0, 5.0]

    walkready:
      travel_time:
        type: double
        description: "Time to walkready in seconds"
        default_value: 5.0
        validation:
          bounds<>: [0.0, 10.0]

  stabilizer:
    end_pause:
      active:
        type: bool
        description: "Stand still until the robot is stable again after standing up"
        default_value: false
      stable_threshold:
        type: double
        description: "Angular threshold for the robot to be considered stable"
        default_value: 0.05
        validation:
          bounds<>: [0.0, 1.0]
      duration:
        type: int
        description: "Number of cycles the robot has to be stable to be considered stable"
        default_value: 100
        validation:
          bounds<>: [0, 1000]
      timeout:
        type: double
        description: "Timeout after which the robot is considered considered stable even if it is not"
        default_value: 2.0
        validation:
          bounds<>: [0.0, 10.0]
    trunk_pid:
      pitch:
        p:
          type: double
          description: "Proportional gain for the trunk pitch PID controller"
          default_value: 0.0
        i:
          type: double
          description: "Integral gain for the trunk pitch PID controller"
          default_value: 0.0
        d:
          type: double
          description: "Derivative gain for the trunk pitch PID controller"
          default_value: 0.0
        i_clamp_max:
          type: double
          description: "Integral clamp maximum for the trunk pitch PID controller"
          default_value: 0.0
        i_clamp_min:
          type: double
          description: "Integral clamp minimum for the trunk pitch PID controller"
          default_value: 0.0
        antiwindup:
          type: bool
          description: "Activate antiwindup for the trunk pitch PID controller"
          default_value: false
      roll:
        p:
          type: double
          description: "Proportional gain for the trunk roll PID controller"
          default_value: 0.0
        i:
          type: double
          description: "Integral gain for the trunk roll PID controller"
          default_value: 0.0
        d:
          type: double
          description: "Derivative gain for the trunk roll PID controller"
          default_value: 0.0
        i_clamp_max:
          type: double
          description: "Integral clamp maximum for the trunk roll PID controller"
          default_value: 0.0
        i_clamp_min:
          type: double
          description: "Integral clamp minimum for the trunk roll PID controller"
          default_value: 0.0
        antiwindup:
          type: bool
          description: "Activate antiwindup for the trunk roll PID controller"
          default_value: false

  visualizer:
    spline_smoothness:
      type: int
      description: ""
      default_value: 100
    display_debug:
      type: bool
      description: ""
      default_value: true

  node:
    tf:
      base_link_frame:
        type: string
        description: "The robots base link frame name"
        read_only: true
        default_value: "base_link"
      r_sole_frame:
        type: string
        description: "The robot right sole frame name"
        read_only: true
        default_value: "r_sole"
      l_sole_frame:
        type: string
        description: "The robot left sole frame name"
        read_only: true
        default_value: "l_sole"
      r_wrist_frame:
        type: string
        description: "The robot right wrist frame name"
        read_only: true
        default_value: "r_wrist"
      l_wrist_frame:
        type: string
        description: "The robot left wrist frame name"
        read_only: true
        default_value: "l_wrist"
